src/run_solver.jl

```diff
@@ -88,27 +94,49 @@ function solve_problems(
   ]
   stats = DataFrame(names .=> [T[] for T in types])
 
+  # Thread-safe mechanisms
+  stats_lock = ReentrantLock()  # Lock for modifying shared data structures
+  first_problem = Atomic{Bool}(true)  # Atomic for safe interaction
+  # specific = Atomic{Vector{Symbol}}([])  # Use atomic for thread-safe updates
   specific = Symbol[]
+  nb_unsuccessful_since_start = Atomic{Int}(0)
 
+  # Prepare DataFrame columns for logging
   col_idx = indexin(colstats, names)
-
-  first_problem = true
-  nb_unsuccessful_since_start = 0
   @info log_header(colstats, types[col_idx], hdr_override = info_hdr_override)
 
-  for (id, problem) in enumerate(problems)
+  # Convert problems to an indexable vector
+  problem_list = collect(problems)
```

Comment:
We really do not want this. If `problems` is a generator of all CUTEst models, this line will try to materialize them all at once and hold them all in memory simultaneously, which is not acceptable.